### PR TITLE
Ranged rogue

### PR DESCRIPTION
**Pull request by catszeid**
Implemented Rogue "range" runes:
- Quick Draw
- Between the Eyes
- Poisoned Knife
- Shuriken Toss

Known issues:

- [x] Shuriken Toss bounces not implemented
- [x] Between the Eyes ap scaling is untested (postponed for testing later)
- [x] Between the Eyes damage variation is untested
- [x] Poisoned Knife can be cast within the 8 yard deadzone (need to look at hunter code)

### CODE REVIEW

File: sim/rogue/between_the_eyes.go [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+	}
+	flatDamage := rogue.RuneAbilityBaseDamage()
+	comboDamageBonus := rogue.RuneAbilityDamagePerCombo()
+	// TODO: Update damage variance
+	damageVariance := map[int32]float64{
+		25: 20,
+		40: 44,
+		50: 68,
+		60: 108,
+	}[rogue.Level]
```
Comment (vigo2) [inline sim/rogue/between_the_eyes.go:22]: Damage variance is always the same - it's from 0.53 to 0.81 multiplied by `flatDamage + cp * comboDamageBonus`.

File: sim/rogue/poisoned_knife.go [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+func (rogue *Rogue) registerPoisonedKnife() {
+	if !rogue.HasRune(proto.RogueRune_RunePoisonedKnife) {
+		return
+	}
+
+	rogue.PoisonedKnife = rogue.RegisterSpell(core.SpellConfig{
+		ActionID:    core.ActionID{SpellID: int32(proto.RogueRune_RunePoisonedKnife)},
+		SpellSchool: core.SpellSchoolPhysical,
+		ProcMask:    core.ProcMaskMeleeOHSpecial,
```
[the remark this answers is not in the document]
Comment (vigo2) [reply sim/rogue/poisoned_knife.go:18]: Nope, it isn't, weirdly...

File: sim/rogue/poisoned_knife.go [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+		ThreatMultiplier: 1,
+		// Cannot Miss
+		BonusHitRating: 100 * core.MeleeHitRatingPerHitChance,
+
+		ApplyEffects: func(sim *core.Simulation, target *core.Unit, spell *core.Spell) {
+			rogue.BreakStealth(sim)
+			baseDamage := spell.Unit.OHNormalizedWeaponDamage(sim, spell.MeleeAttackPower()) + spell.BonusWeaponDamage()
+
+			// Cannot Miss, Dodge, or Parry as per spell flags
+			result := spell.CalcAndDealDamage(sim, target, baseDamage, spell.OutcomeMeleeSpecialNoBlockDodgeParry)
```
Comment (vigo2) [inline sim/rogue/poisoned_knife.go:52]: Nevermind, it's a ranged ability with DefenseType Melee... talk about weird.

https://wago.tools/db2/SpellMisc?build=1.15.1.53181&filter[SpellID]=425012&page=1 - it has "Always hit", "No attack dodge", "No attack parry". Did you ever get _Blocks_ in testing?

It also has pretty much every "suppress proc"-Flag set available, so maybe it doesn't proc at all...

File: sim/rogue/between_the_eyes.go [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+		ApplyEffects: func(sim *core.Simulation, target *core.Unit, spell *core.Spell) {
+			rogue.BreakStealth(sim)
+
+			comboPoints := rogue.ComboPoints()
+			flatBaseDamage := flatDamage + comboDamageBonus*float64(comboPoints)
+
+			// TODO: test combo point AP scaling
+			baseDamage := sim.Roll(flatBaseDamage, flatBaseDamage+damageVariance) +
+				0.03*float64(comboPoints)*spell.MeleeAttackPower() +
```
Comment (vigo2) [inline sim/rogue/between_the_eyes.go:64]: Since it's ranged, it might work with RangedAttackPower

File: sim/rogue/quick_draw.go [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+				Duration: time.Second * 10,
+			},
+			IgnoreHaste: true,
+		},
+		ExtraCastCondition: func(sim *core.Simulation, target *core.Unit) bool {
+			return rogue.Ranged().RangedWeaponType != proto.RangedWeaponType_RangedWeaponTypeThrown ||
+				rogue.Ranged().RangedWeaponType != proto.RangedWeaponType_RangedWeaponTypeUnknown
+		},
+
+		DamageMultiplier: 1 + 0.02*float64(rogue.Talents.Aggression),
```
Comment (vigo2) [inline sim/rogue/quick_draw.go:42]: nit: `[]float64{ .... }` for ~consistency & safety (no "out of range talent points")

File: sim/rogue/quick_draw.go [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+		CritMultiplier:   rogue.RangedCritMultiplier(true),
+		ThreatMultiplier: 1,
+
+		ApplyEffects: func(sim *core.Simulation, target *core.Unit, spell *core.Spell) {
+			rogue.BreakStealth(sim)
+			baseDamage := spell.Unit.RangedWeaponDamage(sim, spell.RangedAttackPower(target)) + spell.BonusWeaponDamage()
```
Comment (vigo2) [inline sim/rogue/quick_draw.go:48]: It does normalized weapon damage (no shortcut, either introduce, or use `rogue.AutoAttacks.ranged.CalculateNormalizedWeaponDamage()`), and I think I remember it also adds Ammo DPS.

**Review by vigo2 — CHANGES_REQUESTED** [review]

File: sim/rogue/poisoned_knife.go [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+				rogue.AddComboPoints(sim, 1, spell.ComboPointMetrics())
+				// 100% application of OH poison (except for 1%? It can resist extremely rarely)
+				if rogue.getImbueProcMask(proto.WeaponImbue_InstantPoison).Matches(core.ProcMaskMeleeOH) {
+					rogue.InstantPoison[ShivProc].Cast(sim, target)
+					return
+				} else if rogue.getImbueProcMask(proto.WeaponImbue_DeadlyPoison).Matches(core.ProcMaskMeleeOH) {
+					rogue.DeadlyPoison[ShivProc].Cast(sim, target)
+					return
+				} else if rogue.getImbueProcMask(proto.WeaponImbue_WoundPoison).Matches(core.ProcMaskMeleeOH) {
+					rogue.WoundPoison[ShivProc].Cast(sim, target)
```
[the remark this answers is not in the document]
Comment (catszeid) [reply sim/rogue/poisoned_knife.go:64]: Poisoned Knife does trigger Instant from Deadly Brew with the normal caveat of needing to roll to hit.

File: sim/rogue/poisoned_knife.go (hunk 3 above)
Comment (catszeid) [reply sim/rogue/poisoned_knife.go:52]: I should test again, but I don't think I saw any Block.

Comment (catszeid) [reply sim/rogue/poisoned_knife.go:52]: I can test for weapon procs later to confirm or deny the behavior

File: sim/rogue/between_the_eyes.go (hunk 1 above)
Comment (catszeid) [reply sim/rogue/between_the_eyes.go]: I updated based on what you said here, but would appreciate a check to make sure I interpreted correctly.

File: sim/rogue/between_the_eyes.go (hunk 4 above)
Comment (catszeid) [reply sim/rogue/between_the_eyes.go:58]: Requires testing. This can be bundled with testing AP scaling (and category of AP)

**Review by vigo2 — APPROVED** [review]
I'll approve, so we don't have to ping-pong again ;)
- BTE damage calculation is still wrong
- Shiv's and Poisoned Knife's Deadly Brew-Instant interaction should be checked

File: sim/rogue/poisoned_knife.go [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+				switch rogue.Consumes.OffHandImbue {
+				case proto.WeaponImbue_InstantPoison:
+					rogue.InstantPoison[ShivProc].Cast(sim, target)
+				case proto.WeaponImbue_DeadlyPoison:
+					rogue.DeadlyPoison[ShivProc].Cast(sim, target)
+				case proto.WeaponImbue_WoundPoison:
+					rogue.WoundPoison[ShivProc].Cast(sim, target)
+				default:
+					if hasDeadlyBrew {
+						rogue.InstantPoison[NormalProc].Cast(sim, target)
```
[the remark this answers is not in the document]
Comment (catszeid) [reply sim/rogue/poisoned_knife.go:67]: As per testing Deadly Brew procs are entirely independent of the trigger source and does not inherit any benefits (such as 100% application).

File: sim/rogue/shiv.go [hunk 9]
```diff
@@ -58,7 +58,7 @@ func (rogue *Rogue) registerShivSpell() {
 					rogue.WoundPoison[ShivProc].Cast(sim, target)
 				default:
 					if hasDeadlyBrew {
-						rogue.InstantPoison[ShivProc].Cast(sim, target)
+						rogue.InstantPoison[NormalProc].Cast(sim, target)
```
[the remark this answers is not in the document]
Comment (catszeid) [reply sim/rogue/shiv.go:61]: Same as above comment, independent proc.

**Pull request merged**